### ISSUE TITLE
Gradients stochasticity study

### ISSUE THREAD

**Issue by Justinezgh** [Justinezgh/SBI-Diff-Simulator#38]
**Our goal is to show that the noisier the gradients, the less they help constrain the distribution.** 
To study this, I created a toy [simulator](https://github.com/Justinezgh/SBI-Diff-Simulator/blob/main/sbids/tasks/toy_model.py). The creation of the simulator is detailed in this [notebook](https://colab.research.google.com/drive/1Bx9zDSrjgPk-ix57PTShHoJaseTP3tcQ?usp=sharing), but briefly, the simulator is built so that the gradients are increasingly noisy with the number of latent variables z. 
The SBI algorithm used here is NPE. The metric used to evaluate the posterior quality (compare to the truth obtained with mcmc) is the [c2st](https://github.com/Justinezgh/SBI-Diff-Simulator/blob/main/sbids/metrics/c2st.py) metric. 

For this experiment I considered two amount of latent variables: 20 and 500. The first part of the experiment was to find the optimal score weight for these two cases:
![image](https://user-images.githubusercontent.com/72011736/220201995-0a636778-c371-4aad-a78d-918d8dde30a1.png)
![image](https://user-images.githubusercontent.com/72011736/220202000-b3f0e4a6-47c6-4d5e-8398-70a37afb736c.png)
I concluded that the optimal score weight for the case with 500 latent variables is 1e-3 and 1e-2 for 20 latent variables. So now we can compare our two cases:
![image](https://user-images.githubusercontent.com/72011736/220202006-056b7864-b906-413a-b83c-0a934f65ba8e.png)
From this plot we can see that in the case of 500 latent variables (so the gradients are noisier than the 20 latent variables case) the quality evaluation curve is above the 20 latent variables one. I think this shows that the noisier the gradients, the less they help constrain the distribution.

(epistemic uncertainty is obtained by averaging 10 NFs)

**Comment by Justinezgh** [comment]
Hi @glouppe I did the experiment again but with different training set for each NF and I got these results: 

![image](https://user-images.githubusercontent.com/72011736/221268441-229295cc-d330-4f18-b478-765669aae45c.png)
![image](https://user-images.githubusercontent.com/72011736/221268459-7c9ab3ac-5cdb-4f84-93a7-258754fce4a8.png)

So it seems that it is not related to the training set. What do you think?
